### PR TITLE
Functional backend refactor (2.x)

### PR DESCRIPTION
**Pull request by jlowin**
Introduces a functional backend for prompt processing. The current backend was designed around function definitions as the primary abstraction (e.g. @fn, @prompt_fn), but as we expand the functional interfaces (e.g. cast, extract), having to go through a function definition is becoming a bottleneck. 

As of now the new backend exists entirely in the v2 submodule, which is just for testing.

### CODE REVIEW

**Comment by zzstoatzz** [comment]
hmm why the switch from `chat.completions` to `llms` in the settings? the first seems more specific / intuitive to me

**Comment by jlowin** [comment]
@zzstoatzz reverted!

File: src/marvin/utilities/python.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+
+class ParameterModel(BaseModel):
+    name: str
+    annotation: Optional[str]
+    default: Optional[str]
+
+
+class PythonFunction(BaseModel, arbitrary_types_allowed=True):
+    function: Callable = Field(description="Original function object")
```
Comment (zzstoatzz) [inline src/marvin/utilities/python.py:17]: could we use `model_config` instead?
```suggestion
from pydantic import ConfigDict
...
class PythonFunction(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)
    function: Callable = Field(description="Original function object")
```

File: src/marvin/requests.py [hunk 2]
```diff
@@ -95,11 +96,53 @@ class ChatRequest(Prompt[T]):
...
     stream: Optional[bool] = False
-    temperature: Optional[Annotated[float, Field(strict=True, ge=0, le=2)]] = 1
+    temperature: Optional[Annotated[float, Field(strict=True, ge=0, le=2)]] = Field(
+        default_factory=lambda: settings.openai.chat.completions.temperature
+    )
     top_p: Optional[Annotated[float, Field(strict=True, ge=0, le=1)]] = 1
     user: Optional[str] = None
 
 
+class ChatResponse(BaseModel, arbitrary_types_allowed=True):
```
Comment (zzstoatzz) [inline src/marvin/requests.py:106]: same as other comment

**Review by zzstoatzz — APPROVED** [review]

**Comment by jlowin** [comment]
Merging with the v2 submodule intact, though note those will ultimately replace the components prior to 2.x release

**Pull request merged**